Prioritized Log Forwarding from the Agent

jasonrclark opened the pull request:
This PR adds prioritized, sampled log collection to the agent. There may be testing loose ends, but the bulk of the work has automated testing already at least partially in place.

This isn't targeted at the main repo's `dev` yet but the branch is usable with Bundler's git targeting if so desired.

## Story of the Changes 📖 
Because it's a lot of changes, I've staged it up to be reviewable by commit. The following will give the high level flow of the changes, and then put some comments inline to highlight points of particular interest throughout.

📈 cce02c152507f0b901fa9a7bf9db42260a2aa083 - We start off **adding performance testing** that we'll want **to baseline before/after** (specifically logging while in transactions; we had the out of transaction case already covered). If we find additional cases to compare they should get rebased back to this commit to make comparison easier.

⛓️  66cbdd38d203cbc21d5344ecca193c446c8b55ae - Logs sent from the agent need the same linking data that we use with the logs-in-context pieces, but the MELT format we'll send them in allows us to avoid repeating the values that are static (i.e. the `entity.guid`, `entity.name` and `entity.type`). To make that easier later, this change **splits out methods for getting the per-transaction linking data from the non-transaction linking data** while keeping the existing bits responding as before.

⏫  e8785fdbdcfcf8861ffbf07c8112ed8e7ca87201 - Next up because it's an easy, standalone case is we **add the logic for calculating a log event's priority**. Note that this algorithm is provisional and up for change, but also very nicely stateless in this module.

🛑 cbf64999a627a3ef1be99c25cb683da8339605ae - **Disable log capture on the `AgentLogger` and `AuditLogger`** classes within the agent. Auditing is never sensible for us to be counting or capturing at all. Agent logs may be desirable but won't be a default, and we have startup ordering issues to make that all correct (currently we only count log lines from the point when the instrumentation is installed, so we miss lots of the agent preamble). Rather than deal with these now, just shut those loggers out. 

🧺 fa8f6d6b66d9a6eb7d2670039394d7fb8259a2e9 - One of the two big pieces -- this **adds the `LogEventAggregator` which is the central class responsible for gathering up logs** until we send them during harvest. This builds off the existing event aggregators, although we do have some bits that are a little different from the others. There may be some refactoring with regards to `EventAggregator` that could happen here. In this commit the class is added, an instance _exists_ on the agent and transactions, but none of the instrumentation sends anything to the aggregator yet.

🐛 482c8918774aa7fcd72d880263615c2883559580 - In adding tests for the `NewRelicService#log_event_data` method I found **we weren't properly testing two other methods on the service** (one wrongly excluded test, one just missing). These methods also weren't making the expected response that other spots were... harmless as the main caller ignores the response, but seemed worth fixing since I was looking at it so they're all the same.

📩 e071b3cfed00a75b53a676fcceaefcb481473ab2 - **Adds the `log_event_data` methods to our services for transmitting** log payloads to the Collector. Note that because of Resque (and maybe other forked processes?) anything that gets on the `NewRelicService` for talking to the Collector _also_ needs to go on the `PipeService` (for talking from a forked process back to the parent process that will talk to NR). These are added, briefly tested but not wired into harvest at this point.

🔑 8f2b888ecd537e38ebb38f05623f2bf11216baf2 - **Adds default for our current (sure to be wrongly named) config keys**. Tests before this were all updated to not assume a given default value. Tests after this may break if we change default to off, but that'll show up immediately.

🎺  74f5684f53132521badb68df23d1e4e4cc4b5e7a - **Instrumentation on `Logger` now hands log events to the aggregator**. The aggregator also takes care now of our metric recording (which we do in batch on each attempted harvest vs every log line which has a huge performance benefit with little downside since even when the aggregator is disabled, the harvest loop still asks each time).

🍞  3b53a190a2327d40d690cdf49909e2e5e7a19c1f - Last and certainly not least, we **wire the aggregator in to send data on each harvest cycle**. The shape of this payload is slightly different than other payloads before but that's mostly isolated in the aggregator.

## Testing 🔬 

Obviously a big consideration on this is the overhead and establishing what reasonable limits are.

Internally there's a test app called log-jam which I've been running these changes on repeatedly as they're under development. This was tested disabling instrumentation, just taking metrics, and with a log buffer of 1000, 2000 and 5000 logs-per-minute. Since the links are all internal I won't put it here, but we may discuss results in the real PR to the main repo.

I'm planning to also get this running on other apps as soon as possible to see real-world impact, as the test app basically _just_ logs, so it's not really representative.

RichVanderwal commented on lib/new_relic/agent/log_priority.rb:
```diff
@@ -0,0 +1,30 @@
+        memo
+      end
+
+      TRANSACTION_BONUS = 10
```
Do we want to break this out even more with a separate value for errors?

jasonrclark replied:
Probably deserves its own constant, even if the values in my current scheme are the same :+1:

jasonrclark committed:
Add perf test for log instrumentation in transactions

jasonrclark committed:
Extract LinkingMetadata for later direct use

jasonrclark committed:
Add class for calculating log priority

jasonrclark committed:
Disable log instrumentation on agent and audit loggers

jasonrclark committed:
Add log aggregator class

jasonrclark committed:
Fix unexpected returns from span and error methods

These don't appear to be _actually_ used in practice
(Agent::Agent#send_data_to_endpoint) but we were testing other methods
to make sure that they conformed to this pattern so worth lining up.

jasonrclark committed:
Add endpoints to service classes for logs

jasonrclark committed:
Add default configuration for log forwarding

These are liable to change before we're done but here they are

jasonrclark committed:
Instrumentation delegates to aggregator

jasonrclark committed:
Wire up harvest to send log event data

jasonrclark commented on lib/new_relic/agent/agent.rb:
```diff
@@ -629,6 +635,10 @@ def create_and_run_event_loop
               transmit_span_event_data
             end
+            @event_loop.on(interval_for CUSTOM_EVENT_DATA) do
+              # TODO: Change to log_event_data when present in collector response
```
As noted in this `TODO` because we don't get a faster harvest response for the log endpoint yet, we're hijacking the one for custom events to get in on the 5-second harvest batching.

kaylareopelle replied:
@jasonrclark  - Is the log_event_data collector response now ready? I wonder if addressing these TODOs should be added the Ruby Agent's MMF plan...

jasonrclark commented on lib/new_relic/agent/configuration/default_source.rb:
```diff
@@ -1839,6 +1839,21 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
           :description => 'Specify a maximum number of custom Insights events to buffer in memory at a time.',
           :dynamic_name => true
         },
+        :'log_sending.enabled' => {
```
This and the following value are very likely to change name but hasn't finalized yet so I'll just touch it once when we get the name.

kaylareopelle replied:
Keep in mind that for the final version, we'll also want to add the settings to `newrelic.yml` so that the community is aware of these options.

jasonrclark replied:
👍  Good call, thanks @kaylareopelle

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+        @counter_lock.synchronize do
+          @seen += 1
+          @seen_by_severity[severity] += 1
+        end
```
These local values get tallied in the aggregator for recording as a metric each harvest. This avoids the expense of coercing metric names in every log call like we're doing in 8.1-8.4.

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+      def record_batch txn, logs
+        # Capture our finalized priority
+        logs.each do |log|
+          log.first["priority"] = LogPriority.priority_for(log.last["level"], txn)
```
TODO: Extract constants

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+            PrioritySampledBuffer::PRIORITY_KEY => priority
+          },
+          event
+        ]
```
This format is slightly funky for our needs but by matching the existing event format internally where it expects to find priority on the events, we can directly use `PrioritySampledBuffer` unchanged for our use case.

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+      end
+
+      # Because our transmission format (MELT) is different than historical
+      # agent payloads, extract the munging here to keep the service focus
```
TODO: Add link to the MELT Log format public docs.

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+      end
+
+      def after_harvest metadata
+        linking = { PLUGIN_TYPE_KEY => PLUGIN_TYPE }
```
TODO: Exclude this if we end up extracting plugin information in the backend

jasonrclark commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,185 @@
+
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Seen", total_count)
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Sent", captured_count)
+        NewRelic::Agent.increment_metric("Supportability/Logging/Customer/Dropped", dropped_count)
```
TODO: Update names once finalized. Also potentially record customer-facing metric for dropped lines.

jasonrclark commented on lib/new_relic/agent/transaction.rb:
```diff
@@ -236,6 +237,7 @@ def initialize(category, options)
 
         @exceptions = {}
         @metrics = TransactionMetrics.new
+        @logs = PrioritySampledBuffer.new(NewRelic::Agent.instance.log_event_aggregator.capacity)
```
Slightly subtle but we cap out each transaction at the per-harvest limit since it can't reasonably keep more than that anyway.

jasonrclark commented on test/multiverse/suites/mongo/mongo2_instrumentation_test.rb:
```diff
@@ -343,15 +343,10 @@ def test_batched_queries
             }
-            assert_metrics_recorded_exclusive expected
+            assert_metrics_recorded_exclusive(expected,
+                                              :ignore_filter => /^(Logging)/)
```
I shoulda used the `:ignore_filter` more broadly the first time around ;)

jasonrclark committed:
Revise awkward common attribute passing

Now that aggregator gets another shot at the payload, don't need to pack
along the linking data the whole time

jasonrclark committed:
Align supportability metrics to spec

jasonrclark committed:
Align existing config names with spec

jasonrclark committed:
Send millis, not seconds

jasonrclark committed:
Revise log priority scheme

jasonrclark commented:
Additions after the initial rebase of this which may be of interest (still commit-by-commit)

🧹  c235467fd0e5fe7d800ef73b6f84d6d4bf036bba - Gets rid of an awkward way the data was passing along metadata that's not really needed after the shape of things changed during development.

📝 56310d966b79ea756f036371bda478e50d4abac9 - Aligning configuration naming to the final choices that were made.

🦸 6ae4383b8ede0fb029d3a353475078332bda51fa - Align metric naming, mostly around supportability values.

🕙 dbbf301bacd7fb58ae217eaef41f2d57d5599739 - Other agents had all agreed they like millis, not seconds for timestamp. Since we can support it with Ruby's `Process.clock_gettime(Process::CLOCK_REALTIME)` being a float and all, we're rolling with it.

jasonrclark committed:
High security mode support

jasonrclark commented:
⏫  b35b0ad8b9fba9c4682c485767ba81e0f0d9e976 - Changed over to the agreed on priority scheme which is 1) take transaction priority 2) lacking that, generate a base priority randomly and 3) first-come-first-serve, no severity ordering.

🔒 44f6a30e873903dc28bb925059c7c0cea9fb63bd - Update so we never gather logs when in high security mode. We will still allow metrics and the eventual local-decoration will be allowed as well, but log forwarding is out.

jasonrclark committed:
Trim unused entity fields from common log payload

jasonrclark commented:
🔪  ec3a4d978f128eef39cd1fa858a16d6c06816215 - Not going to send full `entity.name` and `entity.type` to save volume on the log lines. They aren't required for the in-context features in the UI.

kaylareopelle commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,209 @@
+            end
+          end
+        end
+      rescue
+        nil
```
Why did you decide to rescue using `nil` here?

jasonrclark replied:
Main point here is just to avoid any problem in this code failing the main logging path.

Maybe should be recording something at this point, but typically wouldn't want to return anything accidentally, so putting the `nil` to be explicit about it (rather than whatever our last error handling line happens to return...)

kaylareopelle commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,209 @@
+        nil
+      end
+
+      def record_batch txn, logs
```
JSYK, we've added Rubocop linting to our CI since this PR has been opened, so there may be some small changes to make including putting parenthesis around arguments in a method definition. I believe you're safe to rebase with the `dev` branch, given we've released 8.5.0 without any major issues 🤞

jasonrclark replied:
👍  Once we've done the scheduled review will happily rebase to `dev` and pick that up before we look at actually landing it.

kaylareopelle commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,209 @@
+          return unless @seen > 0
+
+          NewRelic::Agent.increment_metric(LINES, @seen)
+          @seen_by_severity.each do |(severity, count)|
```
This syntax is new to me! Why are parenthesis used within the iterator pipes?

jasonrclark replied:
So that's array destructuring. The `each` on `@seen_by_severity` returns an array of `[key, value]` arrays, and this will put those positional elements into the variables listed in the `()`.

Example pry session (with bonus "remainder" gathering we aren't bothering with here.)

```
[5] pry(main)> pr = Proc.new do |(first, second, *rest)| p first; p second; p rest; end
=> #<Proc:0x00007fcbda141060 (pry):5>

[6] pry(main)> pr.call([1,2,3,4])
1
2
[3, 4]

[7] pry(main)> pr.call([1,2,3])
1
2
[3]

[8] pry(main)> pr.call([1,2])
1
2
[]
```

kaylareopelle commented on test/new_relic/agent/log_event_aggregator_test.rb:
```diff
@@ -0,0 +1,270 @@
+
+    CAPACITY_KEY = LogEventAggregator.capacity_key
+
+    # Helpers for DataContainerTests
```
DataContainerTests are new to me, and I don't believe I quite understand their purpose by reading `test/new_relic/data_container_tests.rb`. What constitutes a DataContainer? When should one be added? (also happy to discuss these during a synchronous meeting if that's easier)

jasonrclark replied:
So a `DataContainer` is the abstraction for "something that buffers data until harvest". The various event aggregators, error collectors and metric takers are all examples.

These "containers" get used somewhat generically by parts of the `NewRelicService` during harvest. The `DataContainerTests` module defines a set of test methods that captures those generic behaviors, largely around required methods, resets and merging items back upon failure.

kaylareopelle reviewed (commented):
Thanks for this solid foundation to add log forwarding to the agent. I have a few comments and also want to leave space for my colleagues to review to see if they unearth other ideas.

kaylareopelle commented on lib/new_relic/agent/log_event_aggregator.rb:
```diff
@@ -0,0 +1,209 @@
+      # Config keys
+      OVERALL_ENABLED_KEY = :'application_logging.enabled'
+      METRICS_ENABLED_KEY = :'application_logging.metrics.enabled'
+      FORWARDING_ENABLED_KEY = enabled_keys.first
```
I see `enabled_keys` and `enabled_key` are aliased. Is the `application_logging.enabled` key not set to these values as a temporary measure? Should it ultimately become the `enabled_key`?

jasonrclark replied:
I hadn't added the overall `application_logging.enabled` key into actual usage (despite having added the constant) so that's part of the weirdness here.

When we wire that up, we can consider whether that's the right thing to use for `enabled_key`. I currently use `enabled?` in the middle of `record` below _after_ metrics but _before_ capturing the log, so that would need to change. Worth evaluating what would make this all the most consistent.

Also of note, my setting `enabled_key :'application_logging.forwarding.enabled'` and then later using `enabled_keys.first` to get at it is dancing around the orphaned config checking, as it recognizes the first form but doesn't follow constants. (Can chat if this doesn't make sense, it's a little weird and subtle).

kaylareopelle commented on test/new_relic/agent/log_event_aggregator_test.rb:
```diff
@@ -0,0 +1,270 @@
+        assert_metrics_recorded_exclusive({
+          "Logging/lines" => { :call_count => 9 },
+          "Logging/lines/DEBUG" => { :call_count => 9 },
+          "Supportability/Logging/Metrics/Ruby/disabled" => { :call_count => 1 },
```
The Engineering Brief currently states the default value for `application_logging.metrics.enabled` is `true`. This would change the metric on line 236 to end with `enabled` instead of `disabled`. Is this accurate for high security mode as well?

jasonrclark replied:
This test isn't quite right because of the unfinished config work we should look at.

Metrics and local decoration configs just don't exist, so that's why they're disabled and 💯  agreed that these metrics should be `enabled`, even under high security when those exist.

The `enabled` on Forwarding below is reading the literal config value vs interpreting it with regards to high security mode, which I _think_ is the intent but maybe deserves double-checking.

jasonrclark commented:
Turned into a real PR over at https://github.com/newrelic/newrelic-ruby-agent/pull/1002
